Duplicate detection

cameron1729 opened the pull request:
Resolves #17

gthomas2 commented on classes/local.php:
```diff
@@ -229,4 +232,22 @@ public static function file_pluginfile($relativepath) {
\ No newline at end of file
+
+    public static function file_is_public(stored_file $file): bool {
+        global $DB;
```
BTW - I'm not asking for a fix to this, just pointing out the implications.

brendanheywood replied:
+1 except I'd go a bit harder and say we need to address this one way or another before this lands. Either caching or something smarted. If we cache then I'd want this to be a localizable cache too

gthomas2 replied:
One possibility would be to have a query which gets all the dupes in one go returning the file hash as the only field. You could cache this entire list of dupes (staticacceleration for speed) and then just use in_array to test for dupes from that point onwards.

gthomas2 commented on tests/filter_test.php:
```diff
@@ -87,7 +87,7 @@ public function test_empty_image() {
         ];
```
I would like a test that ensures urls don't become public if minduplicates is set to 0

brendanheywood commented on classes/image.php:
```diff
@@ -87,10 +88,10 @@ public static function resize (
-        $component = $originalfile->get_component();
-        $filearea = $originalfile->get_filearea();
-        $itemid = $originalfile->get_itemid();
+        $contextid = $public ? 1 : $originalfile->get_contextid();
```
just wondering if this would be clearer as an if else instead of 4 ternaries

gthomas2 replied:
Personally, I think it reads fine as is.

cameron1729 commented:
I believe I've addressed the main changes requested. Additionally, I fixed some cases that the initial patch didn't address:

# When the min duplicates setting is lowered, files that were considered public under the old limit but not public under the new limit were still accessible (if you had the URL). A [change in lib.php](https://github.com/gthomas2/moodle-filter_imageopt/pull/34/commits/df612879a65580dfe64efe17b2a7aa94e4e28f51#diff-8f5ff72ef1cf59199cfc4bbf16f2475d09e448601242f0d2c51bb94bd4aa1395R51) prevents this
# Canonicalisation did not work when images were less than maxwidth. When an image is less than maxwidth but considered public, -1 is used as the width for the purposes of URLs

Finally, the branch has been rebased on work done in the [catalyst repo](https://github.com/catalyst/moodle-filter_imageopt/tree/MOODLE_39_STABLE-catalyst) (which I don't believe had a PR sent). I'm happy to make a separate PR for that one if it feels like too much for this single PR.

@gthomas2 can you please have another look and let me know what you think?

gthomas2 commented on settings.php:
```diff
@@ -49,10 +49,17 @@
-        get_string('widthattributedesc', 'filter_imageopt'), image::WIDTHATTPRSERVELTMAX, $choices));
+        get_string('widthattributedesc', 'filter_imageopt'), local::WIDTH_ATT_PRESERVE_MAX, $choices));
+
+    $settings->add(new admin_setting_configtext(
```
Now we have the cache for the min dupes we also need to purge the cache when this setting changes - see:

$setting->set_updatedcallback

gthomas2 replied:
Ah, I just noticed that the cache is keyed by the minimum dupes setting so you can ignore the above comment.

gthomas2 commented on classes/local.php:
```diff
@@ -229,4 +265,32 @@ public static function file_pluginfile($relativepath) {
+            $publicfilescache->set($key, array_column($publicfiles, 'contenthash'));
+        }
+
+        return in_array($file->get_contenthash(), $publicfilescache->get($key));
```
This is beautiful work. Exactly what I had in mind.

gthomas2 commented on tests/local_test.php:
```diff
@@ -0,0 +1,151 @@
+     * @covers ::file_is_public
+     */
+    public function test_file_is_public() {
+        global $CFG;
```
Excellent work. Exactly what I wanted here.

gthomas2 reviewed (approved):
I'm very happy with this PR and have approved it BUT I would like to give @brendanheywood an opportunity to perform his own review before it is merged. 

Thank you for your awesome work!

brendanheywood commented:
I'm happy with this from a code perspective so +1 from me. There will likely be more patches for new features and more performance tweaks anyway
